article/app/liveblog/KeyEventData.scala

```diff
@@ -0,0 +1,32 @@
+package liveblog
+
+import model.liveblog.BodyBlock.{KeyEvent, SummaryEvent}
+import model.liveblog.{Blocks, BodyBlock, LiveBlogDate}
+import org.scala_tools.time.Imports._
+
+object KeyEventData {
+
+  // just for convenience for use from the templates
+  def apply(maybeBlocks: Option[Blocks], timezone: DateTimeZone): Seq[KeyEventData] = {
+
+    val blocks = maybeBlocks.toSeq.flatMap(blocks => blocks.requestedBodyBlocks.getOrElse(Canonical.timeline, blocks.body))
+
+    apply(blocks, timezone)
+  }
+
+  def apply(blocks: Seq[BodyBlock], timezone: DateTimeZone): Seq[KeyEventData] = {
+
+    val TIMELINE_MAX_ENTRIES = 7
+
+    blocks.foldLeft((false, Nil: List[BodyBlock])) {
+      case ((summaryFound, soFar), nextBlock) if !summaryFound && nextBlock.eventType == SummaryEvent => (true, nextBlock :: soFar)
+      case ((summaryFound, soFar), nextBlock) if nextBlock.eventType == KeyEvent => (summaryFound, nextBlock :: soFar)
+      case ((summaryFound, soFar), _) => (summaryFound, soFar)
+    }._2.reverse.take(TIMELINE_MAX_ENTRIES).map { bodyBlock =>
+      KeyEventData(bodyBlock.id, bodyBlock.publishedCreatedDate(timezone), bodyBlock.title)
+    }
```

Comment:
This foldLeft is clever, but I had to read it a few times before I understood it. It might be kinder to your fellow devs (and your future self) to write it as:

1. Find the latest summary
2. Find the key events
3. Merge them together and take the latest 7

e.g.

```scala
val latestSummary = blocks.find(_.eventType == SummaryEvent)
val keyEvents = blocks.filter(_.eventType == KeyEvent)
(latestSummary.toSeq ++ keyEvents).sortBy(_.lastModified).reverse.take(7)
```